src/psbtv2.ts

```diff
@@ -1026,19 +1057,93 @@ export class PsbtV2 extends PsbtV2Maps {
     this.PSBT_GLOBAL_OUTPUT_COUNT = this.outputMaps.length;
   }
 
+  // The Signer, when it creates a signature, must add the partial sig keypair
+  // to the psbt for the input which it is signing. In the case that a
+  // particular signer does not, this method can be used to add a signature to
+  // the psbt. This method assumes the Signer did the validation outlined in
+  // BIP0174 before creating a signature.
+  // https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#signer
+  public addPartialSig(inputIndex: number, pubkey: Buffer, sig: Buffer) {
+    if (!this.inputMaps[inputIndex]) {
+      throw Error(`PsbtV2 has no input at ${inputIndex}`);
+    }
+
+    if (!pubkey || !sig) {
+      throw Error(
+        `PsbtV2.addPartialSig() missing argument ${
+          (!pubkey && "pubkey") || (!sig && "sig")
+        }`
+      );
+    }
+
+    const key = `${KeyType.PSBT_IN_PARTIAL_SIG}${pubkey.toString("hex")}`;
+    if (this.inputMaps[inputIndex].has(key)) {
+      throw Error(
+        "PsbtV2 already has a signature for this input with this pubkey"
+      );
+    }
+
+    const modBackup = this.PSBT_GLOBAL_TX_MODIFIABLE;
+    try {
+      this.inputMaps[inputIndex].set(key, sig);
+      this.handleSighashType(sig);
+    } catch (err) {
+      console.error(err);
+      // To remain atomic, attempt to reset everything to the way it was.
+      this.inputMaps[inputIndex].delete(key);
+      this.PSBT_GLOBAL_TX_MODIFIABLE = modBackup;
+    }
+  }
+
+  // Used to ensure the PSBT is in the proper state when adding a partial sig
+  // keypair.
+  // https://github.com/bitcoin/bips/blob/master/bip-0370.mediawiki#signer
+  private handleSighashType(sig: Buffer) {
+    const br = new BufferReader(sig.slice(-1));
+    let sighashVal = br.readU8();
+    let modifiable = this.PSBT_GLOBAL_TX_MODIFIABLE;
+
+    if (!(sighashVal & SighashType.SIGHASH_ANYONECANPAY)) {
+      modifiable = modifiable.filter(
+        (val) => val !== PsbtGlobalTxModifiableBits.INPUTS
+      );
+    } else {
+      // Remove SIGHASH_ANYONECANPAY bit for simpler comparisons
```

Comment:
can you explain this comment? I don't really follow.

Reply:
The sighash flag seems to combine and enumeration with a bitmask. If bit 8 is set, it's `SIGHASH_ANYONECANPAY`. The other three flags are enumerated on values of 1, 2, and 3. It was easier to just set bit 8 to 0 and then compare by `value === 3` then to instead have to compare by `value === 3 || value === 131`.

Reply:
ok, gotcha.